hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/lock/OmRequestGatekeeper.java

```diff
@@ -0,0 +1,144 @@
+   * @param lockInfo Defines the locks to acquire.
+   * @return A handle that can be closed to release the acquired locks.
+   * @throws InterruptedException If the thread is interrupted while waiting to acquire some of the locks.
+   * @throws TimeoutException If the timeout elapses while waiting to acquire any of the locks individually.
+   */
+  public AutoCloseable lock(OmLockInfo lockInfo) throws InterruptedException, TimeoutException {
+    // Common case is 1 volume, 1 bucket, and at most 2 key locks.
+    List<Lock> locks = new ArrayList<>(4);
+
+    Optional<LockInfo> optionalVolumeLock = lockInfo.getVolumeLock();
+    Optional<LockInfo> optionalBucketLock = lockInfo.getBucketLock();
+    Optional<Set<LockInfo>> optionalKeyLocks = lockInfo.getKeyLocks();
+
+    if (optionalVolumeLock.isPresent()) {
+      LockInfo volumeLockInfo = optionalVolumeLock.get();
+      if (volumeLockInfo.isWriteLock()) {
+        locks.add(volumeLocks.get(volumeLockInfo.getKey()).writeLock());
+      } else {
+        locks.add(volumeLocks.get(volumeLockInfo.getKey()).readLock());
+      }
+    }
+
+    if (optionalBucketLock.isPresent()) {
+      LockInfo bucketLockInfo = optionalBucketLock.get();
+      if (bucketLockInfo.isWriteLock()) {
+        locks.add(bucketLocks.get(bucketLockInfo.getKey()).writeLock());
+      } else {
+        locks.add(bucketLocks.get(bucketLockInfo.getKey()).readLock());
+      }
+    }
+
+    if (optionalKeyLocks.isPresent()) {
+      for (Lock keyLock: keyLocks.bulkGet(optionalKeyLocks.get())) {
+        locks.add(keyLock);
+      }
+    }
+
+    acquireLocks(locks);
+    return () -> releaseLocks(locks);
+  }
+
+  /*
+  Optional: If we want more diagnostic info on the type of lock that failed to be acquired (volume, bucket, or key),
+  We can make the parameter a list of objects that wrap the Lock with information about its type.
+
+  Note that logging the specific volume, bucket or keys this lock was trying to acquire is not helpful and
+  misleading because collisions within the stripe lock might mean we are blocked on a request for a completely
+  different part of the namespace.
+  Obtaining the thread ID that we were waiting on would be more useful, but there is no easy way to do that.
+   */
+  private void acquireLocks(List<Lock> locks) throws TimeoutException, InterruptedException {
+    List<Lock> acquiredLocks = new ArrayList<>(locks.size());
+    for (Lock lock: locks) {
+      if (lock.tryLock(timeoutMillis, TimeUnit.MILLISECONDS)) {
+        try {
+          acquiredLocks.add(lock);
+        } catch (Throwable e) {
+          // We acquired this lock but were unable to add it to our acquired locks list.
+          lock.unlock();
+          releaseLocks(acquiredLocks);
+          throw e;
+        }
+      } else {
+        releaseLocks(acquiredLocks);
+        throw new TimeoutException("Failed to acquire lock after the given timeout.");
+      }
+    }
+  }
```

Comment:
Concurrently, if the `lock.tryLock(..)` is interrupted, the partial lock won't be released.  Getting everything right with timeout is not easy.